[Priest] rework Shadowflame Prism

seanpeters86 opened the pull request:
issue https://github.com/WarcraftPriests/sl-shadow-priest/issues/77

This adds support for the Shadowflame Prism legendary, along with a few helper functions to address some common use cases.
- Support for `pet.fiend.X` in the APL to reference either Mindbender or Shadowfiend depending on your talent
- Fixes some errors with `base.distance` for Shadow specifically
- Adds `get_current_main_pet()` for checking if the actor is currently using Mindbender or not, this is reused in a few places in the module
- Adds Shadowflame Prism
  - includes a `trigger_shadowflame_prism` method we can call to easy trigger this from Mind Blast and Shadow Word: Death (TODO: maybe we can trigger this from spell data instead of hard coding?)
  - Shadowflame Prism has an impact action for Shadowflame Rift that way we keep the spell data intact
  - Also increases the active pet by 1.5s each time this is called
  - APL line added for SW:D to be used if either bender/shadowfiend is active and you are using the legendary

seanpeters86 committed:
[Priest] rework Shadowflame Prism

seanpeters86 committed:
Merge branch 'shadowlands' into shadowflame-prism

seanpeters86 committed:
try to force type

scamille committed:
[Priest] Add get_current_main_pet function to help with shadowflame prism.

scamille committed:
Pet Spawner: Add missing include

scamille committed:
Pet spawner: Add some more missing includes.

VS was still not compiling.

seanpeters86 committed:
[Priest] finalize shadowflame prism work

scamille commented on engine/class_modules/priest/sc_priest.cpp:
```diff
@@ -2034,12 +2033,24 @@ void priest_t::arise()
   buffs.whispers_of_the_damned->trigger();
 }
 
+void priest_t::trigger_shadowflame_prism( player_t* target )
```
Some things which should probably still be improved here:
1. Check if current_pet is nullptr or not. Either add an assert, or if it is possible to trigger shadowflame prism without having a main pet (eg. no summon lines in the APL) just return without doing anything.

2. Move the assert current_pet->shadowflame_prism inside the !sleeping condition.

3. Why did you remove the set_target line? I would assume it works again after your other fixes.

seanpeters86 committed:
[Priest] make sure we use set_target

scamille commented on engine/class_modules/priest/sc_priest.hpp:
```diff
@@ -983,30 +982,48 @@ struct fiend_melee_t : public priest_pet_melee_t
 struct shadowflame_prism_t final : public priest_pet_spell_t
 {
-  double shadowflame_increase;
+  double duration;
```
please use timespan_t here. That way the units are handled at the point of initialization and not during usage of the value.

scamille commented on engine/class_modules/priest/sc_priest.hpp:
```diff
@@ -983,30 +982,48 @@ struct fiend_melee_t : public priest_pet_melee_t
 
-    set_target( target );
-    execute();
+    auto current_pet = p().o().talents.mindbender->ok() ? p().o().pets.mindbender : p().o().pets.shadowfiend;
```
You should be able to use the get_current_main_pet() function here. Also check if the current_pet variable you retrieve is null or not.

scamille committed:
Priest: Change player distance initialization.

This should set shadow priests correctly to distance of 8 yards, which should then fix shadowcrash usage.

scamille committed:
Priest: Add pet.fiend.X expression

This references either shadowfiend or mindbender

seanpeters86 committed:
[Priest] fix review comments and add apl condition for SWD
